backend/app/user/supabase_user_service.py

```diff
@@ -0,0 +1,158 @@
+from app.models.types import RecordKey
+from app.user import User, UserService
+
+
+class SupabaseUserService(UserService):
+    """
+    A concrete implementation of `UserService`.
+
+    A `SupabaseUserService` uses a Supabase database service to manage the state of a `User`.
+    """
+
+    @override
+    def get_user_anti_recommendations_history(
+        self, user_id: UUID
+    ) -> tuple[RecordKey, ...]:
+        """Return a tuple containing Record keys in a User's anti-recommendation history."""
+
+        try:
+            database_service_result = supabase_database_service.query(
+                TableQuery(
+                    table_name=ARKG_ANTI_RECOMMENDER_USER_STATE_TABLE_NAME,
+                    columns="anti_recommendations_history",
+                    eq={"column": "user_id", "value": str(user_id)},
+                )
+            )
+        except DatabaseException as exception:
+            raise HTTPException(
+                status_code=status.HTTP_400_BAD_REQUEST,
+                detail=f"Unable to get anti-recommendation history of User with id: {user_id!s} from database. Encountered database exception: {exception.message}",
+            ) from exception
+
+        if not database_service_result.succeeded:
+            raise HTTPException(
+                status_code=status.HTTP_400_BAD_REQUEST,
+                detail=f"Unable to get anti-recommendation history of User with id {user_id!s} from database. Check Supabase fetch query parameters",
+            ) from None
+
+        return tuple(database_service_result.data[0]["anti_recommendations_history"])
+
+    @override
+    def get_user_last_seen_anti_recommendation(self, user_id: UUID) -> str:
+        """Return the last Record key in a User's anti-recommendation history."""
+
+        anti_recommendations_history = self.get_user_anti_recommendations_history(
+            user_id
+        )
+
+        if anti_recommendations_history:
+            return anti_recommendations_history[-1]
+
+        return ""
+
+    @override
+    def add_to_user_anti_recommendations_history(
+        self, *, user_id: UUID, anti_recommendation_key: RecordKey
+    ) -> None:
+        """Append `anti_recommendation_key` to a User's anti-recommendation history."""
+
+        try:
+            anti_recommendations_history = list(
+                self.get_user_anti_recommendations_history(user_id)
+            )
+
+            anti_recommendations_history.append(anti_recommendation_key)
+
+            database_service_result = supabase_database_service.command(
+                TableQuery(
+                    table_name=ARKG_ANTI_RECOMMENDER_USER_STATE_TABLE_NAME,
+                    upsert_json={
+                        "anti_recommendations_history": anti_recommendations_history,
+                    },
+                    constraint=str(user_id),
+                )
+            )
+
+        except DatabaseException as exception:
+            raise HTTPException(
```

Comment:
HTTP exceptions don't belong at this level. Throw a UserServiceException or similar and make the HTTP layer deal with it. The service-level exception should have enough information for the HTTP layer to infer the correct response code to send.